specs/executor-economics.md

```diff
@@ -0,0 +1,93 @@
+
+### Executor Selection Strategies
+
+If upgrade execution were to be a free-for-all, we would expect to see Priority Gas Auctions (PGAs) to develop where executors aim to guarantee that their execution transactions land on-chain before the transactions of any other executors. PGAs generally drive up the cost of upgrades and cause executors to pay transaction fees that they cannot be rewarded for. Furthermore, the cost of a PGA is directed to block producers which means that economic value is leaving the ChugSplash system.
+
+To avoid costly PGAs, we introduce the concept of Executor Selection Strategies (ESS). Projects specify a reference to a contract `executor_selection_strategy` which determines the strategy used to select executors for upgrades for the project. It’s the goal of an ESS to minimize inefficiencies as much as possible. Each project can specify its own `executor_selection_strategy`, but executors can choose which strategies they will participate in. By allowing the `executor_selection_strategy` to be a variable configurable by projects, we can introduce opt-in upgrades to newer, more robust ESS contracts.
+
+An ESS contract has the following simple interface:
+
+```solidity
+interface IExecutorSelectionStragegy {
+  /**
+   * Queries the selected executor for a given project/bundle.
+   *
+   * @param _project  Address of the ChugSplashManager that mananges the project.
+   * @param _bundleId ID of the bundle currently being executed.
+   *
+   * @return Address of the selected executor.
+   * @return Time when the executor was selected.
+   */
+  function getSelectedExecutor(
+    address _project,
+    bytes32 _bundleId
+  ) external view returns (address, uint256);
+}
+```
+
+Each ESS then implements its own logic for selecting the executor for a given project and upgrade bundle. A project’s ESS cannot be changed while an upgrade is active. Anyone may create a new ESS at any time.
+
+#### ESS #1: Simple Lock
+
+Our first MVP ESS is the `SimpleLockESS`. Executors may post a bond of `executor_bond_amount` to claim the sole right to execute actions for the locked upgrade for a period of `upgrade_lock_time`. Only the first executor to post a bond gains this right. If the selected executor completes the upgrade within the `upgrade_lock_time` or if the upgrade is cancelled, the bond is returned. If the executor fails to complete the upgrade within the `upgrade_lock_time`, a new executor may post a bond instead and the original bond is forfeited and transferred into the wallet of the project’s `ChugSplashManager` contract. We note that this creates a PGA for the transaction that posts the bond and locks the upgrade but removes PGAs during the execution process.
```

Comment:
> the selected executor completes the upgrade

how is this criteria determined? some bytecode checks onchain?

Reply:
We keep track of the number of executed steps vs the number of total steps. So as soon as executed = total, we say the upgrade is finished.